pkg/util/workloadspread/workloadspread.go

```diff
@@ -61,3 +123,453 @@ func VerifyGroupKind(ref *appsv1alpha1.TargetReference, expectedKind string, exp
+func matchReference(ref *metav1.OwnerReference) (bool, error) {
+	if ref == nil {
+		return false, nil
+	}
+	for _, wl := range workloads {
+		matched, err := VerifyGroupKind(ref, wl.Kind, wl.Groups)
+		if err != nil {
+			return false, err
+		}
+		if matched {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (h *Handler) HandlePodUpdate(oldPod, newPod *corev1.Pod) error {
+	str := newPod.Annotations[MatchedWorkloadSpreadSubsetAnnotations]
+	if str == "" {
+		return nil
+	}
+
+	// Pod has been terminated its lifecycle
+	if oldPod.Status.Phase != newPod.Status.Phase &&
+		(newPod.Status.Phase == corev1.PodFailed || newPod.Status.Phase == corev1.PodSucceeded) {
+		var injectWS *InjectWorkloadSpread
+		err := json.Unmarshal([]byte(str), &injectWS)
+		if err != nil {
+			klog.Errorf("parse Pod (%s/%s) annotations[%s]=%s failed: %s", newPod.Namespace, newPod.Name,
+				MatchedWorkloadSpreadSubsetAnnotations, str, err.Error())
+			return nil
+		}
+		if injectWS == nil || metav1.GetControllerOf(newPod) == nil {
+			return nil
+		}
+
+		matchedWS := &appsv1alpha1.WorkloadSpread{}
+		err = h.Client.Get(context.TODO(), client.ObjectKey{Namespace: newPod.Namespace, Name: injectWS.Name}, matchedWS)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				klog.Warningf("Pod(%s/%s) matched WorkloadSpread(%s) Not Found", newPod.Namespace, newPod.Name, injectWS.Name)
+				return nil
+			}
+			klog.Errorf("get Pod(%s/%s) matched workloadSpread(%s) failed: %s", newPod.Namespace, newPod.Name, injectWS.Name, err.Error())
+			return err
+		}
+
+		return h.mutatingPod(matchedWS, newPod, injectWS, admissionv1beta1.Update)
+	}
+
+	return nil
+}
+
+func (h *Handler) HandlePodCreation(pod *corev1.Pod) error {
+	// filter out pods, include the following:
+	// 1. Deletion pod
+	// 2. Pod.Status.Phase = Succeeded or Failed
+	// 3. Pod.OwnerReference is nil
+	// 4. Pod.OwnerReference is not one of workloads, such as CloneSet, Deployment, ReplicaSet.
+	if !kubecontroller.IsPodActive(pod) {
+		return nil
+	}
+	ref := metav1.GetControllerOf(pod)
+	matched, err := matchReference(ref)
+	if err != nil || !matched {
+		return nil
+	}
+
+	var matchedWS *appsv1alpha1.WorkloadSpread
+	workloadSpreadList := &appsv1alpha1.WorkloadSpreadList{}
+	if err = h.Client.List(context.TODO(), workloadSpreadList, &client.ListOptions{Namespace: pod.Namespace}); err != nil {
+		return err
+	}
+	for _, ws := range workloadSpreadList.Items {
+		if ws.Spec.TargetReference == nil {
+			continue
+		}
+		// determine if the reference of workloadSpread and pod is equal
+		if h.isReferenceEqual(ws.Spec.TargetReference, ref, pod.Namespace) {
+			matchedWS = &ws
+			// pod has at most one matched workloadSpread
+			break
+		}
+	}
+	// not found matched workloadSpread
+	if matchedWS == nil {
+		return nil
+	}
+
+	return h.mutatingPod(matchedWS, pod, nil, admissionv1beta1.Create)
+}
+
+func (h *Handler) HandlePodDeletion(pod *corev1.Pod) error {
```

Comment:
It is recommended that HandlePodDeletion, HandlePodUpdate, HandlePodCreate be merged, in fact there is a lot of logic is similar